citest/json_predicate/path_predicate.py

```diff
@@ -0,0 +1,349 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Support module for locating JSON objects that have certain field values.
```

Comment:
Actually I was not aware of that. I searched for an existing XPath type library when I first started this and didnt come across that. I'll look into supporting that with a variant of this in the future because it has more powerful expression, especially for composed results.

However I am inclined to leave this as is for now because I think being able to see the pruned paths is helpful for debugging. The JsonPath is optimistic, retrieving all the values and is sufficient for what I call candidates. However it wont give insight into the truncated paths so wont help identify why some elements might not have been considered as candidates.

I am open to simplifying things by dropping this. However since it is already done and pretty much encapsulated into this module, I think the value of enriched error reporting and diagnosis for the reader outweighs the complexity of the writer. And this isnt all that bad since the path expression language is very basic.

I'll point out that there is one problem I am aware of here (with what I am doing), and that is that it isnt obvious when to collect a value as a list, and when to enumerate the list and collect its values so I decorate the path with a DONT_ENUMERATE_TERMINAL marker at the end (or really anywhere, but doesnt make sense on an internal node) assuming that you normally want to enumerate the values.